libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,376 @@
+/*
+* AVL Tree implementation.
+* Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+*/
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan {
+
+template<typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable {
+protected:
+    struct Node {
+        T *data;
+        int h = 1; // initially added as leaf
```

Comment:
Reliance of the codebase on native integers like `int` is a mistake that we will be fixing in the rewrite  (non-MISRA-compliant, poor portability, risk of numerical stability issues). Can this be replaced with `uint16_t`?